invisible_cities/reco/deconv_functions_test.py

```diff
@@ -248,10 +297,214 @@ def test_nonexact_binning(data_hdst, data_hdst_deconvolved):
+def test_component_sizes(compsize_array, e_cut, expected_size):
+    '''
+    Given an array, with differing applied e_cuts will
+    provide expected component sizes.
+
+    array = ([[0.5, 0.3, 0.1, 0.3],
+              [0.3, 0.5, 0.1, 0.1],
+              [0.1, 0.3, 0.5, 0.1],
+              [0.1, 0.1, 0.1, 0.5])
+    
+    ecuts:
+    0     ->  array is all 1s, 2 components (zero 0s, sixteen 1s)
+    0.2   ->  array is no longer uniform, 3 components (eight 0s, eight 1s -> two clusters (seven 1s, one 2))
+    0.4   ->  array is identity matrix, 2 components (twelve 0s, four 1s)
+    0.6   ->  array is all 0s, 1 component (sixteen 0s)
+    '''
+
+    bool_mask = np.where(compsize_array < e_cut, 0, 1)
+
+    # check number of components in array match what is expected
+    ccs, no_components = collect_component_sizes(bool_mask)
+    assert(len(no_components) == expected_size)
+
+
+def test_component_elements(compsize_array):
+    '''
+    Given an array, with differing applied e_cuts will
+    provide expected number of elements in each component.
+    
+    The exact logic of the array is described above in test_component_sizes()
+    '''
+    # 0.2 cut, provides 3 components with lengths seven 0s, seven 1s, and one 2
+    e_cut = 0.2
+    
+    # generate labelling
+    bool_mask = np.where(compsize_array < e_cut, 0, 1)
+    ccs, no_components = collect_component_sizes(bool_mask)
+
+    # array to compare against
+    expected_elements = np.array([8, 7, 1])
+    assert(np.allclose(no_components, expected_elements))
+
+    
+
+def test_satellite_zero_protection():
+    '''
+    This test checks that the protection in place to avoid cases in which
+    the number of 0s within the boolean mask from `generate_satellite_mask()` 
+    are less than the satellite_max_size, and so are flagged incorrectly as satellites.
+    '''
+
+    # set size to be just above total number of zeros in given array
+    satellite_max_size = 10
+    
+    e_cut = 0.5
+    cut_type = CutType.abs
+    
+    # provide array with very few 0s that would usually be mislabelled as satellites
+    inverted_array = np.array([[0, 0, 1, 1, 0],
+                               [0, 0, 1, 1, 1],     
+                               [1, 1, 1, 1, 1],     
+                               [1, 1, 1, 0, 0],     
+                               [1, 1, 1, 0, 0]])
+
+    # mask with no satellites, as we don't want zeros to be considered as satellites
+    no_satellite_mask = np.full(inverted_array.shape, False)
+    
+    # generate mask
+    mask = generate_satellite_mask(inverted_array, satellite_max_size, e_cut, cut_type)
+
+    # ensure that no satellites are flagged (as there are none)
+    assert(np.allclose(no_satellite_mask, mask))
+
+def test_generate_satellite_mask_doesnt_modify_input(sat_arr):
+    '''
+    Test that ensures that the applied functions don't modify the provided
+    z-slice array in any way.
+    '''
+    hdst = np.array(sat_arr)
```

Comment:
I notice that you copy the input in every test. It should not be needed. But if it were, you could change `scope="session"` to `scope="function"` in the fixture. This changes the caching behaviour of the fixture.

Reply:
It's because a mask is applied explicitly to the array a few lines further down
```
hdst[mask] = 0
```
Trying to apply this mask explicitly to `sat_arr` worried me due to the scope, but I think it'll be fine to alter the functions such that the mask is applied to `sat_arr` explicitly.

Reply:
Then, what you want to do is to use `scope="function"` in the fixture. This is equivalent to producing a new copy for each test that uses the fixture.